RLogger selection

Axel-Naumann opened the pull request:
```
void func() {
   RLogVerbosityRAII debugThis("FOO", ELogLevel::kDebug);
   Foo::SomethingBroken();
}
```

Axel-Naumann commented:
@phsft-bot build just on mac11.0/default because I changed your config and want to test

Axel-Naumann commented:
@phsft-bot build just on mac11.0/default because I might have fixed your config

eguiraud reviewed (approved):
Niiiice, can't wait to use this :smile:

So I guess the pattern to substitute `if (gDebug > 0) Info(...)` would be

```cpp
// once, e.g. before the event loop, to avoid taking locks in all threads all the time just to
// find out that nothing should be emitted most of the time 
const auto fLogLevel = RLogManager::Get().GetVerbosity("RDF");

// in each thread
if (fLogLevel <= kInfo)
  R__INFO__HERE("RDF") << "doing something in this thread";
```

Does that sound reasonable? Or should we check `RLogManager::Get().GetVerbosity("RDF")` in each TBB task?

More comments inline, all minor.

Axel-Naumann commented:
Will go through your review comments, but first things first: `if (gDebug > 0) printf(...)` is to be replaced by
```
  R__DEBUG_HERE("RDF") << "doing something in this thread";  
```
The fact that you use R__**DEBUG**_HERE signals to `RLogger` that it should only emit this if the verbosity is >= debug.

eguiraud commented:
> The fact that you use R__DEBUG_HERE signals to RLogger that it should only emit this if the verbosity is >= debug.

@Axel-Naumann yes but to figure out whether verbosity (for the group) is >= debug, a "raw" `R__DEBUG_HERE` will take some locks and go through quite some logic won't it?

Axel-Naumann commented:
Sure, but your code does exactly the same, only with more code lines ;-)

eguiraud commented:
I don't think so? I was proposing to retrieve the verbosity level once, before the RDF event loop, and then only call `R__DEBUG_HERE` if the verbosity level requires it (and the check on the cached verbosity level would not lock, so in the common case you'd have no locking). This is at the cost of ignoring changes in the verbosity level that occur during the event loop, of course. I'm open to better solutions :smile:

But I don't think it's reasonable, as a pattern, to unconditionally take locks every time we _might_ want to log something, it would mean that RDF cannot ever put a `R__DEBUG_HERE` or `R__INFO_HERE` in event loop code. `if (gDebug > 0) Info(...)` does not have that issue.

pcanal commented:
> if (gDebug > 0) Info(...) does not have that issue.

To be correct, gDebug (and it's replacement) should be an atomic.

Also (I probably missed something) are we replacing the current multi-level gDebug by a simple bool? (I.e. either nothing or a flood of a data)?

Axel-Naumann commented:
> Also (I probably missed something) are we replacing the current multi-level gDebug by a simple bool? (I.e. either nothing or a flood of a data)?

There are verbosity levels (fatal...debug), both global and per "channel".

eguiraud commented:
> To be correct, gDebug (and it's replacement) should be an atomic.

Depending on the performance impact of checking an atomic in that many places in ROOT, from potentially hundreds of threads, it might be good that gDebug is just an int...

pcanal commented:
> There are verbosity levels (fatal...debug), both global and per "channel".

Yes, so that is a decrease in functionality.  We currently use both the verbosity level (fatal..debug) and for debug with also have a sub-verbosity level (technically 0 through max int)

Axel-Naumann commented:
> so that is a decrease in functionality

Yeah that's something that can be fixed. I'm waiting for your brilliant solution to the issue described by Enrico: how do we deal with thread safety (esp of the channel map) vs speed? A global, thread-racy "nobody wants more verbosity than this"?

pcanal commented:
It is just debug mode, the consequence of a race condition on gDebug would "only" be that some message may or may not be printed when intended.  So it is likely fine.

Axel-Naumann commented:
For @osschar @linev @bellenot : what I want to hear from you is whether the change in log channel name is okay for you:
- "webeve" => "Eve"
- "WebDisplay", "webgui", "CanvasPainter" => "WebGUI"
- "gpadv7", "Gpadv7", "drawable", "Gpad", "rstyle",  => "GPad"
- "Browser7", "rbrowser" => "Browser"

linev commented:
> what I want to hear from you is whether the change in log channel name is okay for you:

Yes, one can change names as you propose. Up to now they was used for error messages, where source of problem can be seen anyway.

Axel-Naumann commented on core/foundation/inc/ROOT/RLogger.hxx:
```diff
@@ -37,7 +37,51 @@ enum class ELogLevel {
+public:
+   RLogEntry() = default;
+   RLogEntry(ELogLevel level, std::string_view group): fGroup(group), fLevel(level) {}
+   RLogEntry(ELogLevel level, std::string_view group, std::string_view filename, int line, std::string_view funcname)
```
Removed `string_view`.

Axel-Naumann commented on core/foundation/inc/ROOT/RLogger.hxx:
```diff
@@ -37,7 +37,51 @@ enum class ELogLevel {
+      : fGroup(group), fFile(filename), fFuncName(funcname), fLine(line), fLevel(level)
+   {}
+
+   RLogEntry &SetFile(const std::string &file)
+   {
+      fFile = file;
```
I decided to not touch it for simplicity...

bellenot commented:
> what I want to hear from you is whether the change in log channel name is okay for you:

Fine with me

Axel-Naumann commented:
I believe I have addressed all your comments:
```
R__DEBUG_HERE(gRDFLog) << "This foo has " << TakesAWhileToCompute() << " bars.";
```
does not take locks and does not call `TakesAWhileToCompute()` if the verbosity (either global or `gRDFLog`'s) are less than `kDebug`. The verbosity comparisons are racy but that's what we want.

pcanal commented on core/foundation/inc/ROOT/RLogger.hxx:
```diff
@@ -82,90 +212,101 @@ public:
    {
-      fFile = file;
-      return *this;
+      return RLogManager::Get().GetVerbosity() >= level || level < ELogLevel::kInfo;
```
How do you suppress Warning message?

Axel-Naumann replied:
`RLogScopedVerbosity silence(ELogLevel::kError)`, unless the channel has warnings turned on. (That's just the answer to your question - we can discuss the interplay of global vs channel verbosity.)

Axel-Naumann replied:
Can you rephrase? I don't understand the second sentence. I'm okay with the first one ;-)

pcanal replied:
Users might be surprised that `RLogManager::Get().SetVerbosity(ELogLevel::kError)` leads to warning message being still printed while `RLogScopedVerbosity silence(ELogLevel::kError)` put in the `main function` removes them.  (Unless I misunderstood something)

Axel-Naumann replied:
`RLogScopedVerbosity silence(ELogLevel::kError)` is just two calls to `RLogManager::Get().SetVerbosity(ELogLevel::kError)`, taking care of resetting to whatever the verbosity was before. I.e.
```
void f() {
  RLogScopedVerbosity silence(ELogLevel::kError);
  Sometihng();
}
```
and
```
void f() {
  auto prev = RLogManager::Get().SetVerbosity(ELogLevel::kError);
  Something();
  RLogManager::Get().SetVerbosity(prev);
}
```
are strictly identical.

Whether a channel's log entries are emitted or not is an "OR" of the global and the channel-specific verbosity: whatever is more verbose wins. Example:

```
RLogChannel ch("Test");
R__WARNING_HERE(ch) << "Emitted";
{
  RLogScopedVerbosity(kError);
  R__WARNING_HERE(ch) << "Suppressed";
}
{
  RLogScopedVerbosity(ch, kError);
  R__WARNING_HERE(ch) << "Emitted"; // global wins - surprisingly?
}
{
  RLogScopedVerbosity(kError);
  RLogScopedVerbosity(ch, kWarning);
  R__WARNING_HERE(ch) << "Visible"; // this time, channel setting wins
}
```

Rationale: allow people to get a silent ROOT - the global verbosity set to `kFatal` should really only show fatal errors unless verbosity is increased for specific channels in which case these channels' logs will be emitted.

One could also argue that a channel's specific setting wins over the global one, and introduce an "unset" channel verbosity, in which case the global verbosity it taken:

```
RLogChannel ch("Test");
R__WARNING_HERE(ch) << "Emitted";
{
  RLogScopedVerbosity(kError);
  R__WARNING_HERE(ch) << "Suppressed";
}
{
  RLogScopedVerbosity(ch, kError);
  R__WARNING_HERE(ch) << "Suppressed"; // channel setting always wins
}
{
  RLogScopedVerbosity(kError);
  RLogScopedVerbosity(ch, kWarning);
  R__WARNING_HERE(ch) << "Visible";
}
```

My feeling was that it might be easier to tell users: "can you run once with `RLogManager::Get().SetVerbosity(ELogLevel::kDebug);`? What's the output?" instead of having to turn on a couple of channels. But I guess that can be done in either scenario. So - shall I change to the second approach because it's less surprising and still allows to do the same things (@pcanal @eguiraud @jblomer mostly, I guess)?

eguiraud replied:
I don't have a lot of experience with logging but "channel-specific wins over global" seems strictly more useful than "more verbose wins" because it _also_ allows running e.g. with "everything on _except_ IO".

However like Philippe in the first comment I'm also confused by the `|| level < ELogLevel::kInfo;`: why do we unconditionally want to create the `RLogEntry` if it's a warning or an error (even if we later won't print them out)? The line probably deserves a comment.

eguiraud commented on core/foundation/src/RLogger.cxx:
```diff
@@ -14,29 +14,33 @@
-bool RLogHandlerDefault::Emit(const ROOT::Experimental::RLogEntry &entry)
+bool RLogHandlerDefault::Emit(const RLogEntry &entry)
 {
    constexpr static std::array<const char *, 5> sTag{{"Debug", "Info", "Warning", "Log", "FATAL"}};
```
```suggestion
   constexpr static std::array<const char *, 5> sTag{{"Debug", "Info", "Warning", "Error", "FATAL"}};
```

eguiraud commented on core/foundation/inc/ROOT/RLogger.hxx:
```diff
@@ -176,13 +317,27 @@ public:
+#define R__NERROR_HERE(CHANNEL, LEVEL) R__LOG_HERE(ROOT::Experimental::ELogLevel::kError + LEVEL, CHANNEL)
+#define R__NWARNING_HERE(CHANNEL, LEVEL) R__LOG_HERE(ROOT::Experimental::ELogLevel::kWarning + LEVEL, CHANNEL)
+#define R__NINFO_HERE(CHANNEL, LEVEL) R__LOG_HERE(ROOT::Experimental::ELogLevel::kInfo + LEVEL, CHANNEL)
+#define R__NDEBUG_HERE(CHANNEL, LEVEL) R__LOG_HERE(ROOT::Experimental::ELogLevel::kDebug + LEVEL, CHANNEL)
```
maybe `OFFSET` rather than `LEVEL`?

Axel-Naumann replied:
`SEVERITY`? ("Offset" to what, and it just says it's a difference but doesn't come with a meaning / "unit")

eguiraud replied:
sure, it was just `LEVEL` that was overloaded

eguiraud reviewed (changes requested):
Nice! It's a "changes requested" just because there are a couple of wrong comments or leftover friend declarations/access specifiers to clean up, the rest is just minor comments.

Axel-Naumann commented on core/foundation/inc/ROOT/RLogger.hxx:
```diff
@@ -16,28 +16,136 @@
+   kError = 0x200,   ///< An error
+   kWarning = 0x300, ///< Warnings about likely unexpected behavior
+   kInfo = 0x400,    ///< Informational messages; used for instance for tracing
+   kDebug = 0x500,   ///< Debug information; only useful for developers
```
Yeah one can now decrease priority of `kError` by making it `kError + 100` or whatever, up to 255. I decided to order them this way around because of "verbosity": more verbose means higher cut-off means `kDebug` must be greater than `kError`. Let me know if you have an opinion on this, i.e. whether it's less surprising to deal with those as "priorities" / "severity" for messages, instead of "verbosity" for filtering. This enum will be used for both, so it's a question of what is more visible / relevant to users. Also @pcanal

Axel-Naumann commented on core/foundation/src/RLogger.cxx:
```diff
@@ -46,14 +50,68 @@ bool RLogHandlerDefault::Emit(const ROOT::Experimental::RLogEntry &entry)
+{
+   std::atomic_init(&fNumWarnings, 0ll);
+   std::atomic_init(&fNumErrors, 0ll);
+   std::atomic_init(&fNumFatalErrors, 0ll);
```
Interesting - that seems to work indeed with all supported compilers. I wonder what made me use `atomic_init`?!

Axel-Naumann commented on core/foundation/inc/ROOT/RLogger.hxx:
```diff
@@ -16,28 +16,136 @@
+ Keep track of emitted errors and warnings.
+ */
+class RLogDiagCount {
+   friend class RLogManager;
```
No because of `++fChannel->fWarning`. I have switched to public methods instead.

eguiraud commented on core/foundation/inc/ROOT/RLogger.hxx:
```diff
@@ -16,28 +16,62 @@
 #include <memory>
+#include <mutex>
 #include <sstream>
 #include <vector>
```
```suggestion
#include <vector>
#include <string>
```

to be very pedantic something for `std::swap` should also be included (and to be even more pedantic something for `std::move` :D )

eguiraud commented on core/foundation/inc/ROOT/RLogger.hxx:
```diff
@@ -82,90 +127,236 @@ public:
+   ELogLevel fVerbosity = ELogLevel::kUnset;
+
+public:
+   /// Construct a throw-away channel, with no channel name and no way of tweaking its verbosity it
```
```suggestion
   /// Construct a throw-away channel, with no channel name and no way of tweaking its verbosity
```

eguiraud commented on core/foundation/inc/ROOT/RLogger.hxx:
```diff
@@ -82,90 +127,236 @@ public:
+
+public:
+   /// Construct a throw-away channel, with no channel name and no way of tweaking its verbosity it
+   /// but with global suppression.
```
> no way of tweaking its verbosity

can't I just call `SetVerbosity` on it?

eguiraud commented on core/foundation/src/RLogger.cxx:
```diff
@@ -14,46 +14,101 @@
+   if (!entry.fLocation.fFuncName.empty())
+      strm << " in " << entry.fLocation.fFuncName;
 
    static constexpr const int errorLevelOld[] = {0, 1000, 2000, 3000, 6000};
```
I don't think these are correct but I might be wrong: we have a `kUnset`, don't have a `kPrint`, `kFatal` is element `1` in our enum but iiuc it's 6000 in `errorLevelOld`...

eguiraud commented on core/foundation/src/RLogger.cxx:
```diff
@@ -14,46 +14,101 @@
+            channel->AddFatalError();
+      }
+   };
+   updateCounts();
```
why do we need a lambda?

eguiraud reviewed (approved):
Sorry for being pedantic :smile: but I bet this stuff will get used _a lot_ for very long.

Looks good to me besides the few comments above. I guess any issue that might remain is minor and can be fixed later.

Axel-Naumann commented:
No worries, thanks for doing that, IMO the code is actually improving by each of your reviews :-)

Axel-Naumann commented:
Needs some more work as suggested by Philippe:
- use namespace-scope functions with local statics instead of global channel variables (as agreed on by @eguiraud but I forgot...);
- fwd declare that channel function and even its return type `RLogChannel` in the relevant headers, to reduce `#include` "load";

And #6931 which will go into a subsequent PR.

eguiraud commented on core/foundation/src/RLogger.cxx:
```diff
@@ -14,46 +14,109 @@
+   if (channel)
+      channel->Increment(entry.fLevel);
+
+   // Is there a sepcific level for the channel? If so, take that,
```
```suggestion
   // Is there a specific level for the channel? If so, take that,
```

eguiraud reviewed (approved):
LGTM!

Axel-Naumann commented on core/foundation/inc/ROOT/RLogger.hxx:
```diff
@@ -55,118 +93,275 @@ public:
+   }
+
+   /// Construct the scoped count for any diagnostic, whatever its channel.
+   RLogScopedDiagCount() : RLogScopedDiagCount(RLogManager::Get()) {}
```
Now sure about this one? I think it's fine to pass `{}` where a `RLogScopedDiagCount` is expected?

eguiraud commented on core/foundation/inc/ROOT/RLogger.hxx:
````diff
@@ -55,118 +93,275 @@ public:
+ ```
+     R__INFO_HERE("CodeChannelForInstanceLibrary") << "All we know is " << 42;
+     const int decreasedInfoLevel = 5;
+     R__NINFO_HERE("CodeChannelForInstanceLibrary", decreasedInfoLevel) << "All we know is " << 42;
````
this is not a thing

Axel-Naumann commented on core/foundation/src/RLogger.cxx:
```diff
@@ -14,46 +14,109 @@
 };
 
-bool RLogHandlerDefault::Emit(const ROOT::Experimental::RLogEntry &entry)
+bool RLogHandlerDefault::Emit(const RLogEntry &entry)
```
Because the class definition would span a 20+ line function - and that's beyond my personal limit for in-class definitions...

Axel-Naumann commented:
I fixed the misconfig of the Mac.

pcanal commented on core/foundation/inc/ROOT/RLogger.hxx:
`````diff
@@ -176,13 +326,43 @@ public:
+ verbosity is too low, i.e.:
+ ````
+ RLogScopedVerbosity silence(RLogLevel::kFatal);
+ R__LOG_DEBUG() << WillNotBeCalled();
`````
This comment/doc might not be matching the code anymore.

Axel-Naumann replied:
Thanks, fixed in 5528d31fbf1a004a580dbe56d85e7a6d077744d7
